src/v/kafka/server/handlers/produce.cc

```diff
@@ -404,6 +404,26 @@ produce_topic(produce_ctx& octx, produce_request::topic& topic) {
             continue;
         }
 
+        // At this point, we know this is an external producer. Reject the write
+        // now if we are critically low on disk space.
+        // XXX TODO performance
+        auto reject = octx.rctx.metadata_cache().should_reject_writes();
```

Comment:
could probably be promoted up out of the produce handler and live in connection_context.cc in a similar way that other quotas are enforced at that higher level.

Reply:
I like this idea.

Reply:
I prototyped this yesterday. The reasons I didn't like it are, roughly:
- Feels like the wrong layer overall.
- It is an awkward spot to generate produce responses, for example.
- We probably still want to keep some statistics on produce errors, which seems harder to do up in the connection ctx layer.
- As we evolve the write rejection logic, we'll need base the decision (reject or not) on things like replica layout, i.e. produce-specific implementation details.

Let me know what you think.